src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -2991,6 +2996,15 @@ void InterpreterSelectQuery::ignoreWithTotals()
     getSelectQuery().group_by_with_totals = false;
 }
 
+bool InterpreterSelectQuery::forceSelectFinalOnSelectQuery(ASTSelectQuery & query)
+{
+    // query.tables() is required because not all queries have tables in it, it could be a function.
+    auto is_force_select_final_setting_on = context->getSettingsRef().force_select_final;
+    auto is_final_supported = storage && storage->supportsFinal() && !storage->isRemote() && query.tables();
```

Comment:
Why not add final for Distributed tables?

Reply:
Distributed tables are supported, I have just added a stateless test as evidence. Please check https://github.com/ClickHouse/ClickHouse/pull/40945/files#diff-637c4a076d097fc4303b8af39daeef42a2b2b0de9bd7394257ef6621194c9361R81-R113